File: Tableau-Supported/Python/insert_data_into_multiple_tables.py
# -----------------------------------------------------------------------------
#
# This file is the copyrighted property of Tableau Software and is protected
# by registered patents and other applicable U.S. and international laws and
# regulations.
#
# You may adapt this file and modify it to fit into your context and use it
# as a template to start your own projects.
#
# -----------------------------------------------------------------------------
from datetime import datetime
from pathlib import Path

from tableauhyperapi import HyperProcess, Telemetry, \
    Connection, CreateMode, \
    NOT_NULLABLE, NULLABLE, SqlType, TableDefinition, \
    Inserter, \
    escape_name, escape_string_literal, \
    HyperException

# Table Definitions required to create tables
orders_table = TableDefinition(
    # Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
    table_name="Orders",
    columns=[
        TableDefinition.Column(name="Address ID", type=SqlType.small_int(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Customer ID", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Order Date", type=SqlType.date(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Order ID", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Ship Date", type=SqlType.date(), nullability=NULLABLE),
        TableDefinition.Column(name="Ship Mode", type=SqlType.text(), nullability=NULLABLE)
    ]
)

customer_table = TableDefinition(
    # Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
    table_name="Customer",
    columns=[
        TableDefinition.Column(name="Customer ID", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Customer Name", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Loyalty Reward Points", type=SqlType.big_int(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Segment", type=SqlType.text(), nullability=NOT_NULLABLE)
    ]
)

products_table = TableDefinition(
    # Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
    table_name="Products",
    columns=[
        TableDefinition.Column(name="Category", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Product ID", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Product Name", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Sub-Category", type=SqlType.text(), nullability=NOT_NULLABLE)
    ]
)

line_items_table = TableDefinition(
    # Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
    table_name="Line Items",
    columns=[
        TableDefinition.Column(name="Line Item ID", type=SqlType.big_int(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Order ID", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Product ID", type=SqlType.text(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Sales", type=SqlType.double(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Quantity", type=SqlType.small_int(), nullability=NOT_NULLABLE),
        TableDefinition.Column(name="Discount", type=SqlType.double(), nullability=NULLABLE),
        TableDefinition.Column(name="Profit", type=SqlType.double(), nullability=NOT_NULLABLE)
    ]
)


def run_insert_data_into_multiple_tables():
    """
    An example of how to create and insert data into a multi-table Hyper file where tables have different types
    """
    print("EXAMPLE - Insert data into multiple tables within a new Hyper file")
    path_to_database = Path("superstore.hyper")

    # Starts the Hyper Process with telemetry enabled to send data to Tableau.
    # To opt out, simply set telemetry=Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU.
    with HyperProcess(telemetry=Telemetry.SEND_USAGE_DATA_TO_TABLEAU) as hyper:

        # Creates new Hyper file "superstore.hyper".
        # Replaces file with CreateMode.CREATE_AND_REPLACE if it already exists.
        with Connection(endpoint=hyper.endpoint,
                        database=path_to_database,
                        create_mode=CreateMode.CREATE_AND_REPLACE) as connection:

            # Create multiple tables.
            connection.catalog.create_table(table_definition=orders_table)
            connection.catalog.create_table(table_definition=customer_table)
            connection.catalog.create_table(table_definition=products_table)
            connection.catalog.create_table(table_definition=line_items_table)

            # Insert data into Orders table.
            orders_data_to_insert = [
                [399, "DK-13375", datetime(2012, 9, 7), "CA-2011-100006", datetime(2012, 9, 13), "Standard Class"],
                [530, "EB-13705", datetime(2012, 7, 8), "CA-2011-100090", datetime(2012, 7, 12), "Standard Class"]
            ]

            with Inserter(connection, orders_table) as inserter:
                inserter.add_rows(rows=orders_data_to_insert)
                inserter.execute()

            # Insert data into Customers table.
            customer_data_to_insert = [
                ["DK-13375", "Dennis Kane", 518, "Consumer"],
                ["EB-13705", "Ed Braxton", 815, "Corporate"]
            ]

            with Inserter(connection, customer_table) as inserter:
                inserter.add_rows(rows=customer_data_to_insert)
                inserter.execute()

            # Insert individual row into Product table.
            with Inserter(connection, products_table) as inserter:
                inserter.add_row(row=["TEC-PH-10002075", "Technology", "Phones", "AT&T EL51110 DECT"])
                inserter.execute()

            # Insert data into Line Items table.
            line_items_data_to_insert = [
                [2718, "CA-2011-100006", "TEC-PH-10002075", 377.97, 3, 0.0, 109.6113],
                [2719, "CA-2011-100090", "TEC-PH-10002075", 377.97, 3, None, 109.6113]
            ]

            with Inserter(connection, line_items_table) as inserter:
                inserter.add_rows(rows=line_items_data_to_insert)
                inserter.execute()

            tables = [orders_table, customer_table, products_table, line_items_table]
            for table in tables:
                # `execute_scalar_query` is for executing a query that returns exactly one row with one column.
                row_count = connection.execute_scalar_query(query=f"SELECT COUNT(*) FROM {table.table_name}")
                print(f"The number of rows in table {table.table_name} is {row_count}.")

        print("The connection to the Hyper file has been closed.")
    print("The Hyper process has been shut down.")


if __name__ == '__main__':
    try:
        run_insert_data_into_multiple_tables()
    except HyperException as ex:
        print(ex)
        exit(1)

File: Tableau-Supported/CPP/insert_data_into_multiple_tables.cpp
// -----------------------------------------------------------------------------
//
// This file is the copyrighted property of Tableau Software and is protected
// by registered patents and other applicable U.S. and international laws and
// regulations.
//
// You may adapt this file and modify it to fit into your context and use it
// as a template to start your own projects.
//
// -----------------------------------------------------------------------------

/**
 * \example insert_data_into_multiple_tables.cpp
 *
 * An example of how to create and insert into a multi-table Hyper file with different column types.
 */

#include <hyperapi/hyperapi.hpp>
#include <iostream>
#include <string>

/** Table Definitions required to create tables */
using Column = hyperapi::TableDefinition::Column;
static const hyperapi::TableDefinition ordersTable{"Orders", // Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
                                                   {
                                                      Column{"Address ID", hyperapi::SqlType::smallInt(), hyperapi::Nullability::NotNullable},
                                                      Column{"Customer ID", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                      Column{"Order Date", hyperapi::SqlType::date(), hyperapi::Nullability::NotNullable},
                                                      Column{"Order ID", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                      Column{"Ship Date", hyperapi::SqlType::date(), hyperapi::Nullability::Nullable},
                                                      Column{"Ship Mode", hyperapi::SqlType::text(), hyperapi::Nullability::Nullable},
                                                   }};
static const hyperapi::TableDefinition customerTable{"Customer", // Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
                                                     {Column{"Customer ID", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                      Column{"Customer Name", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                      Column{"Loyalty Reward Points", hyperapi::SqlType::bigInt(), hyperapi::Nullability::NotNullable},
                                                      Column{"Segment", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable}}};
static const hyperapi::TableDefinition productTable{"Products", // Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
                                                    {
                                                       Column{"Category", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                       Column{"Product ID", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                       Column{"Product Name", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                       Column{"Sub-Category", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                    }};
static const hyperapi::TableDefinition lineItemsTable{"Line Items", // Since the table name is not prefixed with an explicit schema name, the table will reside in the default "public" namespace.
                                                      {
                                                         Column{"Line Item ID", hyperapi::SqlType::bigInt(), hyperapi::Nullability::NotNullable},
                                                         Column{"Order ID", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                         Column{"Product ID", hyperapi::SqlType::text(), hyperapi::Nullability::NotNullable},
                                                         Column{"Sales", hyperapi::SqlType::doublePrecision(), hyperapi::Nullability::NotNullable},
                                                         Column{"Quantity", hyperapi::SqlType::smallInt(), hyperapi::Nullability::NotNullable},
                                                         Column{"Discount", hyperapi::SqlType::doublePrecision(), hyperapi::Nullability::Nullable},
                                                         Column{"Profit", hyperapi::SqlType::doublePrecision(), hyperapi::Nullability::NotNullable},
                                                      }};

static void runInsertDataIntoMultipleTables() {
   std::cout << "EXAMPLE - Insert data into multiple tables within a new Hyper file" << std::endl;
   const std::string pathToDatabase = "data/superstore.hyper";

   // Starts the Hyper Process with telemetry enabled to send data to Tableau.
   // To opt out, simply set telemetry=hyperapi::Telemetry::DoNotSendUsageDataToTableau.
   {
      hyperapi::HyperProcess hyper(hyperapi::Telemetry::SendUsageDataToTableau);
      // Creates new Hyper file "superstore.hyper"
      // Replaces existing file with hyperapi::CreateMode::CreateAndReplace if it already exists.
      {
         hyperapi::Connection connection(hyper.getEndpoint(), pathToDatabase, hyperapi::CreateMode::CreateAndReplace);
         const hyperapi::Catalog& catalog = connection.getCatalog();

         // Create multiple tables.
         catalog.createTable(ordersTable);
         catalog.createTable(customerTable);
         catalog.createTable(productTable);
         catalog.createTable(lineItemsTable);

         // Insert data into Orders table.
         {
            hyperapi::Inserter inserter(connection, ordersTable);
            inserter.addRow(static_cast<int16_t>(399), "DK-13375", hyperapi::Date{2012, 9, 7}, "CA-2011-100006", hyperapi::Date{2012, 9, 13}, "Standard Class");
            inserter.addRow(static_cast<int16_t>(530), "EB-13705", hyperapi::Date{2012, 7, 8}, "CA-2011-100090", hyperapi::Date{2012, 7, 12}, "Standard Class");
            inserter.execute();
         }

         // Insert data into Customer table.
         {
            hyperapi::Inserter inserter(connection, customerTable);
            inserter.addRow("DK-13375", "Dennis Kane", 518, "Consumer");
            inserter.addRow("EB-13705", "Ed Braxton", 815, "Corporate");
            inserter.execute();
         }

         // Insert individual row into Product table.
         {
            hyperapi::Inserter inserter(connection, productTable);
            inserter.addRow("TEC-PH-10002075", "Technology", "Phones", "AT&T EL51110 DECT");
            inserter.execute();
         }

         // Insert data into Line Items table.
         {
            hyperapi::Inserter inserter(connection, lineItemsTable);
            inserter.addRow(2718, "CA-2011-100006", "TEC-PH-10002075", 377.97, int16_t{3}, 0.0, 109.6113);
            inserter.addRow(2719, "CA-2011-100090", "TEC-PH-10002075", 377.97, int16_t{3}, hyperapi::null, 109.6113);
            inserter.execute();
         }

         for (auto& tableName : {ordersTable.getTableName(), customerTable.getTableName(), productTable.getTableName(), lineItemsTable.getTableName()}) {
            // `executeScalarQuery` is for executing a query that returns exactly one row with one column.
            int64_t rowCount = connection.executeScalarQuery<int64_t>("SELECT COUNT(*) FROM " + tableName.toString());
            std::cout << "The number of rows in table " << tableName << " is " << rowCount << "." << std::endl;
         }
      }
      std::cout << "The connection to the Hyper file has been closed." << std::endl;
   }
   std::cout << "The Hyper Process has been shut down." << std::endl;
}

int main() {
   try {
      runInsertDataIntoMultipleTables();
   } catch (const hyperapi::HyperException& e) {
      std::cout << e.toString() << std::endl;
      return 1;
   }
   return 0;
}
